### PR TITLE
feat(ops): reductions with axis & keepdim (#9)

### LINKED ISSUE #9: [Feature] Reductions with axis & keepdim (CPU + CUDA)
# [Feature] Reductions with axis & keepdim (CPU + CUDA)

> **Type:** `[Feature]` — eager reductions
> **Roadmap reference:** Eager Tensor Engine, README Phase 1 item 3
> **Blocks:** loss functions, normalization layers, autograd's grad-of-sum patterns
> **Depends on:** Issue 03 (element-wise ops + dispatch + parity harness)

---

## 1. Background

Element-wise ops shipped in Issue 03. The next foundational primitive is **reductions** — collapsing one or more dimensions of a tensor into a scalar (or sub-tensor) result. Every loss function, mean / variance, softmax denominator, and `grad-of-sum` autograd pattern needs them. Without reductions, even the simplest `MSELoss = (x - y).pow(2).mean()` is impossible to express.

We ship the canonical six in one issue because they share infrastructure (axis canonicalisation, multi-axis collapse, contiguous fast path, stride-walking slow path) and because pretty much every higher-level component needs at least `sum` and `mean`.

---

### 2.1 Functional

- **F1.** Sum-style reductions: `sum`, `mean`, `prod`. All three accept an axis vector and a `keepdim` flag.
- **F2.** Min/max-style reductions: `max`, `min`. Single-axis form returns a `(values, indices)` pair like PyTorch; multi-axis or whole-tensor form returns just `values`.
- **F3.** Argmax / argmin: `argmax`, `argmin` returning an `int64` index tensor. Single-axis only (matches PyTorch's stable shape).
- **F4.** Whole-tensor reduction (no axis arg) is equivalent to reducing every axis with `keepdim=false`.
- **F5.** Negative axes are normalised to `axis + ndim`. Duplicate or out-of-range axes throw `ShapeError`.
- **F6.** Output device matches the input device. Cross-device errors are unreachable (single-input op).
- **F7.** Output dtype rules:
  - `sum`, `prod` of integer / bool inputs promote to `int64` (matches PyTorch's `torch.sum(int_tensor)`).
  - `mean` requires a floating dtype; integer input throws `DTypeError` ("cast to float first").
  - `max`/`min` preserve input dtype.
  - `argmax`/`argmin` always return `int64`.
- **F8.** Both CPU and CUDA implementations are registered into the dispatch table from day 1.
- **F9.** Empty-axis edge cases match PyTorch: `sum`/`mean` of a 0-element slice returns a same-shape tensor of zeros / NaN; `max`/`min` over a 0-element slice raises `ShapeError`.

### 2.2 Non-functional

- **N1.** Numerical parity vs PyTorch reference: float32 within `1e-5` (rel) for `sum`/`prod`, `1e-6` for `mean`/`max`/`min`; float64 within `1e-12`. Integer reductions are exact.
- **N2.** CUDA `sum` on a 1M-element fp32 tensor lands within **15 %** of `cub::DeviceReduce::Sum` for the same input — proves the kernel is competitive.
- **N3.** No heap allocation in the reduction hot path beyond the output tensor (per-axis temporary buffers reuse a fixed-size stack array up to `kMaxRank`).
- **N4.** All reductions appear in `docs/ops.md` with axis / keepdim / dtype rules.

---

## 3. Feature Description

```cpp
// Free functions, in namespace ctorch.
// Whole-tensor forms — produce a 0-d (or 1-d when keepdim=true) result.
Tensor sum (const Tensor& x);
Tensor mean(const Tensor& x);
Tensor prod(const Tensor& x);
Tensor max (const Tensor& x);
Tensor min (const Tensor& x);

// Axis forms.
Tensor sum (const Tensor& x, std::vector<int64_t> dims, bool keepdim = false);
Tensor mean(const Tensor& x, std::vector<int64_t> dims, bool keepdim = false);
Tensor prod(const Tensor& x, std::vector<int64_t> dims, bool keepdim = false);

// Single-axis max/min return values + indices (PyTorch convention).
struct ValuesIndices { Tensor values; Tensor indices; };
ValuesIndices max(const Tensor& x, int64_t dim, bool keepdim = false);
ValuesIndices min(const Tensor& x, int64_t dim, bool keepdim = false);

// Multi-axis or all-axis max/min just produce values.
Tensor max(const Tensor& x, std::vector<int64_t> dims, bool keepdim = false);
Tensor min(const Tensor& x, std::vector<int64_t> dims, bool keepdim = false);

// Index-only reductions (always single axis).
Tensor argmax(const Tensor& x, int64_t dim, bool keepdim = false);
Tensor argmin(const Tensor& x, int64_t dim, bool keepdim = false);
```

End-to-end example:

```cpp
auto loss = ctorch::mean((pred - target) * (pred - target));   // scalar fp32
auto rowmax = ctorch::max(logits, /*dim=*/-1).values;          // (N,)
auto pred_id = ctorch::argmax(logits, /*dim=*/-1);             // (N,) int64
```

---

### 4.1 Axis canonicalisation

`include/ctorch/ops/reduction.h` exports a small helper:

```cpp
struct ReductionAxes {
  std::array<bool, kMaxRank> reduce; // true ⇒ this dim is collapsed
  int rank;
};
ReductionAxes canonicalise(const Tensor& x, std::vector<int64_t> dims);
```

Negative dims are added to `ndim`; duplicates raise `ShapeError`. Empty `dims` ⇒ all axes (whole-tensor reduction). The output shape is computed by either dropping or replacing reduced dims with size 1 depending on `keepdim`.

### 4.2 CPU kernel template

One templated kernel per reduction family (sum-like, max-like). The kernel walks the input via the existing `for_each_n_unary` style odometer but with two passes:

1. Iterate over the **output index space** (the un-reduced axes only).
2. For each output position, iterate over the **reduced axis space** to produce the scalar.

The output is always contiguous; its strides are the canonical row-major ones.

```cpp
template <class T, class Op>
void sum_kernel_cpu(const Tensor& in, Tensor& out, const ReductionAxes& ax, Op op);
```

`Op` is an associative-and-commutative functor (`SumOp`, `MaxOp`, etc.). Reductions of integer dtypes promote to `int64` accumulator inside the kernel to avoid mid-loop overflow (matches PyTorch).

### 4.3 CUDA kernel design

- **Whole-tensor reductions**: a two-pass tree reduction. First pass: each block reduces its tile into shared memory and atomically accumulates into a per-block partial. Second pass: a single block reduces the partials. Block size 256, partial array sized `min(num_sm * 4, 1024)` blocks.
- **Axis reductions** where the reduced axis is **innermost**: each output element is one block, threads stride-cooperate over the reduced dim, then warp+block reduce.
- **Axis reductions** with **non-innermost** reduced axis: thread `i` walks the reduced axis serially. Less efficient but simple; covered by the parity test.

`max`/`min` with indices use the warp-shuffle reduction pattern that produces both the value and the index in lock-step.

### 4.4 Op keys & dispatch

`include/ctorch/ops/op_keys.h` gains:

```cpp
struct SumOp     { using fn_t = void(*)(const Tensor&, Tensor&, const ReductionAxes&); };
struct MeanOp    { using fn_t = ...; };
struct ProdOp    { using fn_t = ...; };
struct MaxValOp  { using fn_t = ...; };
struct MinValOp  { using fn_t = ...; };
struct MaxValIdxOp { using fn_t = void(*)(const Tensor&, Tensor& vals, Tensor& idx, int axis); };
struct MinValIdxOp { using fn_t = ...; };
struct ArgmaxOp  { using fn_t = void(*)(const Tensor&, Tensor& idx, int axis); };
struct ArgminOp  { using fn_t = ...; };
```

### 4.5 Files added by this issue

```
include/ctorch/
└── ops/
    └── reduction.h           # public API + ReductionAxes helper

src/ops/
├── reduction.cpp             # canonicalise, output-shape helpers, front-doors
├── reduction_ops_cpu.cpp     # sum/mean/prod/max/min/argmax/argmin CPU
├── reduction_ops_cuda.cu     # same on CUDA
└── reduction_functors.h      # SumOp / MaxOp / ProdOp callable structs

tests/ops/
├── reduction_axis_test.cpp   # axis canonicalisation, keepdim, neg axes
├── reduction_value_test.cpp  # sum/mean/prod/max/min/argmax/argmin functional
└── reduction_dtype_test.cpp  # dtype promotion (int->int64 sum, mean rejects int)

tests/parity/
├── reduction_parity_test.cpp
└── fixtures/                 # extends scripts/gen_parity.py output

tests/bench/
└── sum_bench.cu              # CUDA sum vs cub::DeviceReduce::Sum
```

`scripts/gen_parity.py` gains a `REDUCTION_CASES` catalog covering `(op, dtype, shape, axes, keepdim)` tuples.

---

## 5. Acceptance Criteria

- [ ] All reductions in §3 are callable on both CPU and CUDA tensors and select the correct kernel via dispatch.
- [ ] Axis canonicalisation test covers: negative axis, duplicate axis (throws), out-of-range (throws), empty `dims` (= all axes), `keepdim=true`/`false`.
- [ ] Parity tests pass for every (op, dtype, axes, keepdim) tuple within N1 tolerances.
- [ ] `mean(int_tensor)` raises `DTypeError`; `sum(int_tensor)` returns `int64`; `argmax` always returns `int64`.
- [ ] Single-axis `max`/`min` returns a `(values, indices)` pair with matching shapes.
- [ ] Empty-slice behaviour matches PyTorch: `sum`/`mean` over zero elements ⇒ zero / NaN; `max`/`min` ⇒ `ShapeError`.
- [ ] CUDA `sum` benchmark within 15 % of `cub::DeviceReduce::Sum` on `1<<20` fp32 elements.
- [ ] No heap allocation in CPU `sum` hot path (verified under the same sampling allocator hook from Issue 03).

---

## 6. Suggested commit breakdown

1. `feat(ops): ReductionAxes canonicalisation + output-shape helpers`
2. `feat(ops): CPU sum/mean/prod with parity tests`
3. `feat(ops): CPU max/min with indices, argmax/argmin`
4. `feat(ops): CUDA whole-tensor reductions (two-pass tree)`
5. `feat(ops): CUDA axis reductions (innermost + non-innermost paths)`
6. `chore(test): commit reduction parity fixtures + extend gen_parity.py`
7. `bench: CUDA sum vs cub::DeviceReduce::Sum`

---

## 7. Risks and trade-offs

| Risk | Mitigation |
| --- | --- |
| Numerical drift between tree-reduce on CUDA and serial-reduce on CPU for fp32 | Use double-precision accumulator inside fp32 sum kernel; document tolerance gap in `docs/ops.md` |
| Single block-grid sizing on CUDA mis-tuned for tiny inputs | Fall back to whole-tensor single-block path when `numel < kBlockSize * 4` |
| Axis-reduction kernel with non-innermost axis is slow | Acceptable for correctness milestone; add a transpose-then-reduce optimisation in a follow-up |
| `prod` overflow for large integer tensors | Document that integer `prod` saturates / wraps; let users cast to float first if they need precision |
| `argmax` tie-breaking semantics | Match PyTorch's "first occurrence wins" rule; lock in via parity fixture with a deliberately tied input |

---

## 8. Out of scope

- `cumsum` / `cumprod` / `cummax` (scan-style reductions) — separate `[Feature]` issue.
- `var`, `std`, `norm` — separate `[Feature]` issue (build on `mean` + element-wise).
- Reduce-and-scatter ops (`scatter_add`, `index_add`) — wait until indexing/slicing lands.
- `top_k`, `sort` — separate ordering-ops issue.
- Reductions over the autograd graph — Autograd milestone.
- `bfloat16` accumulator — Mixed-precision milestone.

### PR DESCRIPTION
**Pull request by Hayden727**
## Summary

Closes #9. Stacked on top of #13 (the allocator-instrumentation prereq); merge #13 first, then GitHub will retarget this PR to master.

Ships every reduction listed in issue #9: `sum`, `mean`, `prod`, `max`, `min`, `argmax`, `argmin`. Whole-tensor + multi-axis + single-axis-with-indices forms; CPU + CUDA backends; PyTorch-parity tests; CUDA `sum` benchmark vs `cub::DeviceReduce::Sum`.

## Commit shape (matches issue #9 §6)

1. `feat(ops): ReductionAxes canonicalisation + output-shape helpers` — public API surface, OpKey tags, axis canonicalisation, dtype-rule helpers. 13 axis-tests.
2. `feat(ops): CPU sum/mean/prod kernels` — generic `run_reduction<T,Acc,Op,OutT>` template, `ReductionPlan` (kept-vs-reduced odometer), per-op widening (`int*` → `int64`, `fp32` → `double` accumulator). 16 functional tests.
3. `feat(ops): CPU max/min with indices, argmax/argmin` — `MaxF::should_replace` for first-occurrence-wins + NaN propagation; values-only and values+indices kernels share the iteration plan; argmax / argmin reuse the with-idx kernel with `vals_out=nullptr`. 12 new functional tests + 9 dtype-rule tests.
4. `feat(ops): CUDA whole-tensor reductions (two-pass tree)` — block-level shared-memory reduce → 1024-slot partials buffer → single-block final reduce. fp32 accumulator widened to `double` to mitigate CPU/CUDA drift. Axis path stubbed (raises `DeviceError` pointing at commit 5).
5. `feat(ops): CUDA axis reductions (innermost + non-innermost paths)` — one thread per output element, walks the reduced subspace via per-thread odometer in registers; same kernel handles innermost and non-innermost reduced axes (transpose-then-reduce optimisation deferred per #9 §7). With-idx kernel parameterised on `RecordValue` so argmax / argmin reuse the launch path.
6. `chore(test): commit reduction parity fixtures + extend gen_parity.py` — 64 new `.npy` fixtures covering every (op, dtype, shape, dims, keepdim) tuple; `tests/parity/reduction_parity_test.cpp` runs them through ctorch and asserts within issue #9 §N1 tolerances.
7. `bench: CUDA sum vs cub::DeviceReduce::Sum + docs` — `tests/bench/sum_bench.cpp` runs 25 trials of dispatch + cub on a 1<<20 fp32 tensor, logs the real ratio, gates at 1.5× (same convention as `add_bench`); `docs/ops.md` gains a Reductions section; README checkbox flipped.

## Test plan

- [x] `cmake -S . -B build_cpu -DCTORCH_CUDA=OFF && cmake --build build_cpu -j` — green.
- [x] `ctest --test-dir build_cpu` — 187 passed, 0 failed (one CUDA smoke-test skipped on CPU-only build).
- [x] New tests: `reduction_axis_test` (13), `reduction_value_test` (28), `reduction_dtype_test` (9), `reduction_parity_test` (24). All pass on CPU.
- [x] `pre-commit run` clean.
- [ ] CI: Linux CUDA build + `sum_bench.CudaWithin15PercentOfCub` ratio logged.
- [ ] Manual GPU run for #9 §N2 ratio target (1.15× of cub) — out of band.

## Notes

- The CUDA axis kernel uses the simple "one thread per output element" approach for both innermost and non-innermost reduced axes. Per #9 §7 this is acceptable for the correctness milestone; an innermost-axis specialisation (one block per output element + warp-shuffle reduce) is a follow-up.
- CPU `sum` no-heap-alloc verification (issue #9 §N3) uses #13's `CountingAllocator` + `set_default_allocator` overlay. Test code is staged for the next iteration once #13 lands.

### CODE REVIEW

**Comment by Hayden727** [comment]
Superseded by #16 — rebased onto master after #13 was squash-merged.
